### ISSUE TITLE
NonCopyable structs attribute and analyzer

### ISSUE THREAD

**Issue by jkotas** [dotnet/runtime#50389]
## Background and Motivation

New high-performance APIs are often exposed as structs to avoid GC heap allocation. Callers of such APIs have to be careful to avoid creating accidental copies of the struct. Failure to do so can lead to correctness or security issues. We need a capability to prevent or detect this class of bugs at build time.

`ValueStringBuilder` is an example of an API where creating accidental copy is a potential security bug: https://github.com/dotnet/runtime/issues/25587#issuecomment-662076796

Existing implementations of similar analyzers:
- https://github.com/ufcpp/NonCopyableAnalyzer
- https://github.com/dotnet/roslyn-analyzers/pull/3420

## Proposed API

Promote https://github.com/ufcpp/NonCopyableAnalyzer developed by @ufcpp into .NET platform API so that it can be used by the platform itself.

```diff
 namespace System
 {
+    [AttributeUsage(AttributeTargets.Struct)]
+    public class NonCopyableAttribute : Attribute
+    {
+    }
 }
```

## Usage Examples

```csharp
[NonCopyable]
public struct ValueStringBuilder
{
...
}

ValueStringBuilder vsb = new ValueStringBuilder();
f(vsb); // Error. ValueStringBuilder must by passed by reference
```

## Alternative Designs

Full C# language feature. The difficulty in doing so is described in https://blog.paranoidcoding.com/2019/12/02/borrowing.html .

## Risks

Corner cases that are missed by the analyzer, e.g. use of reflection.

**Comment by jkotas** [comment]
More context https://github.com/dotnet/designs/pull/189#discussion_r594683843
cc @GrabYourPitchforks

**Comment by sharwell** [comment]
The one feature that I really wish the analyzer supported is "allow copies to read-only locations". This configuration would allow non-mutating operations to run against snapshot copies, which broadens applicability of `[NonCopyable]` from _strictly_ non-copyable types to more general usage on mutable value types to avoid mutating a copy that will not be observed.

**Comment by jkotas** [comment]
> allow copies to read-only location

It would not work for `ValueStringBuilder` and other similar cases where the struct holds manually managed resource.

**Comment by bartonjs** [comment]
> The one feature that I really wish the analyzer supported is "allow copies to read-only locations"

That wouldn't solve the ValueStringBuilder problem, though.  At least, not if it uses ArrayPool.

* A VSB gets into a given state.
* A readonly copy gets created for the current state.
* The writable version calls Append
  * VSB needs to grow, so it rents a new array, copies the contents, returns the old array.
* State sanity check, the copy still looks like it did prior to the call to Append.
* Something else happens
  * The something else rents an array, gets the recently returned array, and uses it.
* State sanity check fails.
  * The copy had a reference to the returned and re-rented array.  That array's contents have changed.

Even without ArrayPool, overwrites would be visible on the copy (writing through into arrays), but appends wouldn't (changing fields).

I don't doubt that readonly copies can make sense in some cases, but they wouldn't solve this one, and feel like they're a weird knifes-edge in general.

**Comment by sharwell** [comment]
> It would not work for `ValueStringBuilder` and other similar cases where the struct holds manually managed resource.

Note that my comment was more for wishing we could do this:

```
[NonCopyable(AllowCopyToReadOnlyLocation = true)]
```

Types like `ValueStringBuilder` would use the default value `false`.

**Comment by sharwell** [comment]
On a totally separate note, this feature is going to be extremely difficult to implement without the compiler exposing some sort of an API specifically to analyze value copies.

**Comment by tannergooding** [comment]
CC. @jaredpar. Would this conflict with or be superseded by any work being done on the language side?

**Comment by sharwell** [comment]
I'm not aware of LDM work on this, but will let @jaredpar provide a more definitive answer. I stopped pushing on this primarily due to questions surrounding the value of https://github.com/dotnet/runtime/issues/50389#issuecomment-809662254 (the compiler would not be able to provide this value through built-in language features, but an analyzer could account for it as long as an analyzer API for value copies existed).

**Comment by jaredpar** [comment]
Coulpe of questions:

- Do you all have a fleshed out set of rules for `NonCopyable` or `ValueStringBuilder` that you want to enforce with this analyzer? Couple that seem to apply:
    - Disallowed as generic arguments
    - Disallow `[NonCopyable] struct` as a field of standard `struct`
    - Disallow as locals in an `async` method or iterator
    - etc ...
- Generally non-copyable types end up desiring to have move semantics. Essentially copying is bad but move is okay. Do you all desire this property here and if so how do you all define what constitutes a "move". Consider that returning a local from a function is likely a safe "move" operation. 
- Are multiple mutable references okay here? Basically can you do the following when `vsb` represents a `[NonCopyable] struct`: `SomeMethod(ref vsb, ref vsb)`. 
- Are `[NonCopyable]` types forced to be `ref struct` or can they be simple `struct`? If the latter then that implies tearing is okay (is that the case)? 

> On a totally separate note, this feature is going to be extremely difficult to implement without the compiler exposing some sort of an API specifically to analyze value copies.

I agree. There are a lot of subtle cases where the compiler does or does not copy a value. It's possible to write conservative analyzers which flag more copies than exist but are unlikely to miss any. It's hard to write one that exactly matches the behavior of the compiler though. 

It's also important to know if we care about invisible copies. There are several places where the compiler will introduce copies in synthesized members that are invisible to the customer but do constitute copies of the values (consider the places where the compiler introduces thunks to meet some metadata contract and hence has to pass values through the thunk). Do this matter for this feature? My assumption is no but wanted to check.

**Comment by sharwell** [comment]
> Disallowed as generic arguments

This is still a known hole in RS0042

> Disallow [NonCopyable] struct as a field of standard struct

This was one of the holes we recently fixed in RS0042.

> Disallow as locals in an async method or iterator

This should be safe right?

> etc ...

Boxing is the main one. Originally I wrote RS0042 to disallow boxing, but later I realized that boxing is a safe option (when it's moved to a box) and _unboxing_ is the problematic operation. A boxed value type can be accessed using `Unsafe.Unbox` or through a virtual call.

> Generally non-copyable types end up desiring to have move semantics. Essentially copying is bad but move is okay. Do you all desire this property here and if so how do you all define what constitutes a "move". Consider that returning a local from a function is likely a safe "move" operation.

For all the cases I've encountered to date, implicit move semantics were desirable (i.e. a copy is only a problem if the location is read after that copy). If a type is not movable, there are many additional constraints (e.g. cannot store as a field of a reference type unless that reference type is pinned, and cannot capture a local into a reference type).

> Are multiple mutable references okay here?

💭 This _should_ be a safe operation.

> Are [NonCopyable] types forced to be ref struct or can they be simple struct?

For the types I've been working with, making the types a `ref struct` is extremely undesirable. This breaks the ability to use the types in state machines.

**Comment by Joe4evr** [comment]
Also related: https://github.com/dotnet/csharplang/issues/2372

**Comment by Joe4evr** [comment]
> Disallowed as generic arguments

I imagine an exception could be designed here, similar to how people want an exception to passing `ref struct`s as a type argument:
```cs
public void M<[NonCopyableSafe] T>(ref T val)
{
    //compiler checks that all uses of T don't make "forbidden" copies, then callers are allowed to substitute a non-copyable struct type
} 
```

Of course, this can always be done later and isn't strictly necessary for an MVP. I'm just throwing some 🍝.

**Comment by huoyaoyuan** [comment]
Related: https://github.com/dotnet/csharplang/discussions/4345

**Comment by stephentoub** [comment]
@333fred, how would this play with how you're defining builders for interpolated strings?  Presumably some builders (e.g. ones wrapping ArrayPool buffers) would benefit from being annotated as non-copyable, but the design as proposed today actually requires them to be passed around.

**Comment by jkotas** [comment]
> Would this work for e.g. SpinLock

Yes. More examples of NonCopyable types in Roslyn: https://github.com/dotnet/roslyn/search?q=%5BNonCopyable%5D

**Comment by GrabYourPitchforks** [comment]
@jaredpar I tried to draft a list of restrictions a while back (see https://gist.github.com/GrabYourPitchforks/7ab6a440100467a82cfe5998cd1e91be). I saw Jan linked to this gist in an earlier comment, but I don't know how much it influenced this proposal or whether there was any kind of value judgment made over it. When you and I talked about this a few weeks back you had pointed out some rough spots.

**Comment by 333fred** [comment]
> @333fred, how would this play with how you're defining builders for interpolated strings? Presumably some builders (e.g. ones wrapping ArrayPool buffers) would benefit from being annotated as non-copyable, but the design as proposed today actually requires them to be passed around.

The answer for that will depend on what LDM things about builders passed by `in` or ref, I guess.

**Comment by stephentoub** [comment]
> The answer for that will depend on what LDM thinks about builders passed by in or ref, I guess.

Yes... we should have an opinion / proposal, though.

**Comment by jaredpar** [comment]
> > The answer for that will depend on what LDM thinks about builders passed by in or ref, I guess.
> Yes... we should have an opinion / proposal, though.

I actually think this is a good motivating example for why we need to think about "move" semantics. 

Consider that `in` is unlikely to be the answer here. The builders need to be mutated in the calling function, if for nothing else than to call `ToString`. That is a mutating function for the builders, `ValueStringBuilder` returns items to the array pool, hence it's unlikely `ToString` will be `readonly`. That means it won't be callable on `in` unless we first copy which violates `NonCopyable`. 

Passing by `ref` will work but the design specifically allows for multiple builders to be in play here. That means we're going to have cases where multiple builders are being passed by `ref` and the builders are likely to be `ref struct`. That gets into territory where it gets easy to run afowl of `ref struct` lifetime rules.

The idea of "moving" the value, essentially passing it to the function such that the calling function's copy is destroyed (assigned `default`) seems like it is the most friction free option.

**Comment by benaadams** [comment]
> The idea of "moving" the value, essentially passing it to the function such that the calling function's copy is destroyed (assigned `default`) seems like it is the most friction free option.

That would mean it couldn't be used for `SpinLock` as that would also destroy its locking semantics as copying would.

**Comment by jkotas** [comment]
> That would mean it couldn't be used for SpinLock

We may need multiple levels:
- No copy/move (e.g. `SpinLock`)
- Move is ok (e.g. `ValueStringBuilder`)
- Read-only copy is ok (e.g. https://github.com/dotnet/runtime/issues/50389#issuecomment-809662254)

**Comment by acaly** [comment]
I have had a closely related issue. I sometimes internally use a mutable struct to implement "fast list", which is just a wrapper of array, from which references of the elements can be obtained. I remember once I marked such a struct with readonly. As a result, adding elements fails because it cannot modify the field to the newly allocated array in case of initialization and reallocation.

So in some cases, if not all, I would suggest to have a way to declare that a struct should not be used as readonly field. I guess many _move-is-ok_ noncopyable structs should fit in this category.

This is more like a language feature instead of runtime feature.

**Comment by sharwell** [comment]
@benaadams @jkotas Can you explain why move semantics would not work for `SpinLock`? Looking at the implementation, it appears that move semantics would be fine.

**Comment by acaly** [comment]
@sharwell I guess it's because someone might be waiting for the lock, effectively storing a reference to the struct on stack.

**Comment by sharwell** [comment]
@stephentoub A move is defined as the relocation of a value from location A to a different location B, after which point no code will access location A expecting to observe the original value. Atomicity of the move is not guaranteed; code which accesses the location is expected to either provide synchronization for the move or intentionally operate on a value which can be moved atomically.

**Comment by stephentoub** [comment]
So how are you envisioning that would work with a SpinLock:
```C#
private SpinLock _lock;
```
What code do two threads write to successfully use that lock?  Today it's:
```C#
_lock.Enter();
...
_lock.Exit();
```
What is it when SpinLock is attributed as being moveable?

**Comment by sharwell** [comment]
It depends entirely on the manner in which the instance is moved. For the case of `StrongBox<SpinLock>`, there is no special handling needed for the case where the compacting GC moves the storage location for the boxed value.

**Comment by stephentoub** [comment]
SpinLock is rarely used that way.  It's typically a field in and of itself.  So how does it work in that case?

**Comment by sharwell** [comment]
Non-copyable value types are types where the user's code is responsible for managing the storage location for the value. Likewise, moving a non-copyable value type is an operation implemented in the user's code, with all the caveats faced by the compacting GC. Application misbehavior following failure to perform a move with correct semantics does not mean a correctly-performed move operation would have the same result.

**Comment by sharwell** [comment]
Put another way: do we have examples of situations where a user is likely to misuse a `SpinLock` without a warning if we implement it as non-copyable but movable?

**Comment by stephentoub** [comment]
Sam and I spoke offline.  The disconnect stems from two completely different definitions of moveable:
- I was using it to mean an API / language syntax for copying the value and clearing out the original or otherwise invalidating the original.
- Sam was using it to mean that its address in memory could change, which would be the case for anything on the GC heap.

**Comment by danmoseley** [comment]
@stephentoub thoughts about whether we should aim to do work here in 7.0 (either analyzer or language feature)? Seems there's a fair bit of interest in ValueStringBuilder alone.

**Comment by stephentoub** [comment]
> @stephentoub thoughts about whether we should aim to do work here in 7.0 (either analyzer or language feature)?

In theory it's a valuable thing to do and would unblock a set of features we've been hesitant to do, making them enough less of a footgun that we'd then move forward with them.  In practice the initial work associated with this would be coming up with the actual rules it would implement, and then seeing which of the things we've avoided doing would now be practical/still valuable given those rules.  As is my typical want, we'd need to be ready to use it in multiple places in the same release that we add it.  That goes whether it's done as a C# language/compiler or analyzer feature.  We'd also need to work through any compat concerns with applying the attribute to types we've already shipped, and what level of noise we'd be ok with on existing code.

**Comment by mrpmorris** [comment]
I've been stung by this twice with `SpinLock`

First
```c#
public readonly SpinLock Locker = new();
```

Second
```c#
public static void ExecuteLocked(this SpinLock locker)
{
  // Above should have `ref`
}
```

**Comment by sschoener** [comment]
Hello! I realize that this issue has been silent for a while, but I wanted to add some thoughts.

In the Unity game engine on the C# side, we had structs with relative pointers embedded. We eventually added an analyzer to detect and disallow value copies for these structs specifically. However, this is insufficient to get an acceptable developer experience: In the case of relative pointers, copying the struct is not just a a security risk or a performance issue but a correctness problem. All C# debuggers that I have tried actively copy structs around. In the best case you get invalid data in the debugger, in the worst case you get debugger crashes. I have written about this [here](https://blog.s-schoener.com/2024-12-17-unity-blobs/), for the curious.

Maybe this particular problem is mostly a consequence of using unsafe code, but at the very least it stands as a reminder that "thou shalt not copy structs" may have implications for tooling that treats value types generically, like a debugger.

**Comment by JeremyKuhne** [comment]
@jaredpar, @phil-allen-msft  Is this the right place to be tracking adding this feature or the equivalent for C#? It would be really good to be able to start productizing things like `ValueStringBuilder`. #25587 The only thing I'm finding is https://github.com/dotnet/roslyn/issues/26937.

**Comment by jaredpar** [comment]
I _think_ that runtime is the right place. In terms of language we have a _very rough_ idea about what a _move only_ style solution would look like. But this is a feature that would be driven by runtime need. Specifically need the full set of motivating examples and push from the runtime team on a release that we'd target for this.

**Comment by tannergooding** [comment]
Closing this as it is highly unlikely to be exposed or supported as an analyzer.

Rather, this is much more likely to be driven by a language feature; particularly related to potential future ownership rules/semantics to help mitigate known memory safety considerations.

The discussion will persist here for historical context and any future design work can be driven off a dotnet/csharplang proposal or discussion thread.
